### DISCUSSION TITLE
Access outages due to renewal procedure for access tokens

### DISCUSSION THREAD

**Discussion by lgblaumeiser** [eclipse-tractusx/tractusx-edc#1203] [Ideas]
For the Golden Record Use Case, we face the issue at Cofinity-X, that our customers complain about the reaction time delay of our service due to the token renewal done by the EDC after the access token for an existing contract expires. The renewal process takes a significant amount in time, in which the service is unavailable. The preferred way would be that there is always a valid token, so that at all times, accessing the Golden Record data is possible without a delay. 

A possible idea that came up in our company would be that the EDC does the token renewal just prior to the expiration time, so that the refreshed token is already available when the old token expires.

 What could we do to prevent these outtages and provide a smooth service experience?

**Comment by paullatzelsperger** [comment]
I'm assuming you are referring to versions >= `0.7.0-RCx`, and if so, then this is taken care of by DataPlane Signaling, which has a built-in token renewal mechanism, c.f. [documentation](https://github.com/eclipse-tractusx/tractusx-edc/tree/main/docs/development/dataplane-signaling)

**Comment by lgblaumeiser** [comment]
Thanks @paullatzelsperger , one last question, can the client find out locally, when the token expires? I assume, that the token contains a expiration time stamp field, which can be checked against the current time, right? In this case, the client could easily implement a timer, that executes the token refresh so, that the available token is always valid.

**Reply by paullatzelsperger** [reply]
yes, the Endpoint Data Reference (i.e. the data structure that consumers will receive) will contain a field named `expiresIn` (cf. [Tractus-X Profiles](https://github.com/eclipse-tractusx/tractusx-profiles/blob/main/tx/refresh/refresh.token.grant.profile.md)), which contains the expiry in seconds (as per OAuth2 standard). Note that this is a _relative_ timestamp that has to be "added to the current timestamp at the time of message reception".

**Reply by lgblaumeiser** [reply]
Thanks, this should be sufficient for the use case.
